framework/src/main/java/org/checkerframework/framework/ajava/InsertAjavaAnnotations.java

```diff
@@ -0,0 +1,583 @@
+        if (!existingImports.contains(importString)) {
+          newImports.add(importString);
+        }
+      }
+
+      if (!newImports.isEmpty()) {
+        int position;
+        int lineBreaksBeforeFirstImport;
+        if (!dest.getImports().isEmpty()) {
+          Position lastImportPosition =
+              dest.getImports().get(dest.getImports().size() - 1).getEnd().get();
+          position = getFilePosition(lastImportPosition) + 1;
+          lineBreaksBeforeFirstImport = 1;
+        } else if (dest.getPackageDeclaration().isPresent()) {
+          Position packagePosition = dest.getPackageDeclaration().get().getEnd().get();
+          position = getFilePosition(packagePosition) + 1;
+          lineBreaksBeforeFirstImport = 2;
+        } else {
+          position = 0;
+          lineBreaksBeforeFirstImport = 0;
+        }
+
+        String insertionContent = "";
+        // In Java 11, use String::repeat.
+        for (int i = 0; i < lineBreaksBeforeFirstImport; i++) {
+          insertionContent += System.lineSeparator();
+        }
+        insertionContent += String.join("", newImports);
+
+        insertions.add(new Insertion(position, insertionContent));
+      }
+
+      src.getModule().ifPresent(l -> l.accept(this, dest.getModule().get()));
+      src.getPackageDeclaration()
+          .ifPresent(l -> l.accept(this, dest.getPackageDeclaration().get()));
+      for (int i = 0; i < src.getTypes().size(); i++) {
+        src.getTypes().get(i).accept(this, dest.getTypes().get(i));
+      }
+    }
+
+    /**
+     * Creates an insertion for a collection of annotations. The annotations will appear on their
+     * own line (unless any non-whitespace characters precede the insertion position on its own
+     * line).
+     *
+     * @param position the position of the insertion
+     * @param annotations List of annotations to insert
+     */
+    private void addAnnotationOnOwnLine(Position position, List<AnnotationExpr> annotations) {
+      String line = lines.get(position.line - 1);
+      int insertionColumn = position.column - 1;
+      boolean ownLine = true;
+      for (int i = 0; i < insertionColumn; i++) {
+        if (line.charAt(i) != ' ' && line.charAt(i) != '\t') {
+          ownLine = false;
+          break;
+        }
+      }
+
+      if (ownLine) {
+        StringJoiner insertionContent = new StringJoiner(" ");
+        for (AnnotationExpr annotation : annotations) {
+          insertionContent.add(printer.print(annotation));
+        }
+
+        if (insertionContent.length() == 0) {
+          return;
+        }
+
+        String whitespaceCopy = line.substring(0, position.column - 1);
+        int filePosition = getFilePosition(position);
+        insertions.add(
+            new Insertion(
+                filePosition,
+                insertionContent.toString() + System.lineSeparator() + whitespaceCopy,
+                true));
+      } else {
+        addAnnotations(position, annotations, 0, false);
```

Comment:
I think the way it is now makes the most sense since we usually don't want to add extra space before an annotation. For example, in a method declaration `void myMethod(@Tainted String s1, @Untainted String s2)`, each annotation has a space before it if and only if there was a space before the original type it was attached to, that is, there is no extra space before the annotations. Are you saying we should treat the annotation differently since it was requested to have its own line?

Reply:
The `addAnnotations` method on this line already adds a space after the insertion content. The whitespace that precedes an inserted annotation should already be present before its insertion. Setting the last parameter to `true`, I think, is only necessary in very niche edge cases in array types.

Reply:
OK.  Thanks for the explanation!  And thanks again for this program.